### ISSUE TITLE
Question about Conv-layer and Conv_Block

### ISSUE THREAD

**Issue by jihwan722** [yaorong0921/Driver-Intention-Prediction#12]
We are running main_inside.py and main_outside.py, and I want to know if the content related to Conv_layer in main_outside.py is the same as the picture.

![image](https://github.com/yaorong0921/Driver-Intention-Prediction/assets/127845632/cf6c1518-37bc-493a-9fa9-8be85282c729)

If not, achieving a size of 64×37×59 after passing through 3x112x176 is impossible with 3x3 conv layer with stride 1 and 2x2 Max pool with stride 2.

Please provide an answer. Thank you.

**Comment by jihwan722** [comment]
I mean conv_layers in the classifier in the below picture
![image](https://github.com/yaorong0921/Driver-Intention-Prediction/assets/127845632/e28c5f5c-2faf-4743-880d-c11e5030783d)
Is that in the code you provided?

**Comment by yaorong0921** [comment]
Dear @jihwan722,
Before the decoder, there is a convLSTM encoder. Table II refers to the output after the encoder. Did you include it in the pipeline?

**Comment by jihwan722** [comment]
I mean achieving a size of 64×37×59(Conv Block 0's output size) after passing through 32x112x176 is impossible with 3x3 conv layer with stride 1 and 2x2 Max pool with stride 2. Please provide your answer for this.

**Comment by yaorong0921** [comment]
@jihwan722 , please use a 3x3 conv layer with stride 3 to shrink the input size from 112x176 to 37x59 first. Code about classifier is not included in main_outside.py.

**Comment by jihwan722** [comment]
Thanks for your reply!!
I want to ask more about fusion two modules which are inside module and outside module. 

First, the inside images are 25fps and outside images are 30fps. 
Plus, when we make trainloader we use training_data from dataset.py, but training_data_inside's has 1 n_samples per each video and training_data_outside's has 10 n_samples per each video.

So, we want to fusion two modules but number of data is different.

Can we get the classifier module code? gns453@kookmin.ac.kr

How can we combine two moduels?

**Comment by jihwan722** [comment]
`import os
import sys
import json
import torch
import torch.nn as nn
from opts import parse_opts
from model import generate_model
from dataset import get_training_set_inside,get_training_set_outside
from spatial_transforms import (
    Compose, Normalize, Scale, CenterCrop, CornerCrop, MultiScaleCornerCrop,
    MultiScaleRandomCrop, RandomHorizontalFlip, ToTensor, DriverFocusCrop, DriverCenterCrop)
from temporal_transforms import LoopPadding, TemporalRandomCrop, TemporalCenterCrop, UniformRandomSample, UniformEndSample, UniformIntervalCrop

from target_transforms import ClassLabel
from torch.autograd import Variable
from mean import get_mean, get_std
from models.convolution_lstm import encoder, classifier
from utils import AverageMeter, calculate_accuracy


def extract_inside_vector(inside_model_path, input_data):
    
    model, _ = generate_model(opt)
    checkpoint = torch.load(inside_model_path)
    model.load_state_dict(checkpoint['state_dict'])
    
    model.eval()
    with torch.no_grad():
        inputs = input_data.float().to(device)
        outputs = model(inputs)
        inside_vector = outputs.cpu().numpy()

    return inside_vector

def extract_outside_vector(outside_model_path, input_data):
    model = encoder(hidden_channels=[128, 64, 64, 32], sample_size=opt.sample_size, sample_duration=opt.sample_duration_outside).to(device)    
    checkpoint = torch.load(outside_model_path)
    state_dict = checkpoint['state_dict'] 
    new_state_dict = {key.replace('module.', ''): value for key, value in state_dict.items()}
    model.load_state_dict(new_state_dict)

    model.eval()
    with torch.no_grad():
        inputs = input_data.float().to(device)
        outputs = model(inputs)
        # To Tensor
        outside_vector = outputs.cpu().numpy()

    return outside_vector

class Conv_Block(nn.Module):
    def __init__(self):
        super(Conv_Block, self).__init__()

        self.conv_block = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=0),
            # nn.BatchNorm2d(64),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2, padding=0)
            
        )
        self.conv_block2 = nn.Sequential(
            nn.Conv2d(64, 128, kernel_size=3, stride=1, padding=0),
            # nn.BatchNorm2d(128),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2, padding=0)
            
        )
        self.conv_block3 = nn.Sequential(
            nn.Conv2d(128, 256, kernel_size=3, stride=1, padding=0),
            # nn.BatchNorm2d(256),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2, padding=0)
            
        )
        self.conv_block4 = nn.Sequential(
            nn.Conv2d(256, 512, kernel_size=3, stride=1, padding=0),
            # nn.BatchNorm2d(512),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2, padding=0)
            
        )
        self.conv_block5 = nn.Sequential(
            nn.Conv2d(512, 512, kernel_size=3, stride=1, padding=0),
            # nn.BatchNorm2d(512),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=3, stride=3, padding=0)
            
        )

    def forward(self, x):
        x = self.conv_block(x)
        x = self.conv_block2(x)
        x = self.conv_block3(x)
        x = self.conv_block4(x)
        x = self.conv_block5(x)
        #flatten the output
        x = x.view(x.size(0),-1)
        return x
    
class classifier(nn.Module):
    def __init__(self):
        super(classifier, self).__init__()
        # FC0: 3072 -> 2048
        
        self.classifier_fc = nn.Sequential(
            nn.Linear(3072, 2048),
            # nn.BatchNorm1d(2048),
            nn.ReLU(),
            nn.Linear(2048, 5),
            # nn.BatchNorm1d(5),
            nn.ReLU(),
            nn.Softmax(dim=1) 
        )

    def forward(self, x):
        x = self.classifier_fc(x)
        return x


if __name__ == '__main__':
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
    inside_pth_path = './inresults/save_best.pth'
    outside_pth_path = './outresults/convlstm-save_best.pth'
    
    
    # inside 훈련 데이터 증강 및 데이터 로더 부분
    opt = parse_opts()
    
 

    opt.scales_inside = [opt.initial_scale]
    for i in range(1, opt.n_scales_inside):
        opt.scales_inside.append(opt.scales_inside[-1] * opt.scale_step)
    opt.arch_inside = '{}-{}'.format(opt.model, opt.model_depth)
    opt.mean_inside = get_mean(opt.norm_value_inside, dataset=opt.mean_dataset)
    opt.std_inside = get_std(opt.norm_value_inside)
    

    torch.manual_seed(opt.manual_seed)

    
    if opt.no_mean_norm and not opt.std_norm:
            norm_method = Normalize([0, 0, 0], [1, 1, 1])
    elif not opt.std_norm:
        norm_method = Normalize(opt.mean_inside, [1, 1, 1])
    else:
        norm_method = Normalize(opt.mean_inside, opt.std_inside)
        
        
        
    if not opt.no_train_inside:
        assert opt.train_crop in ['random', 'corner', 'center', 'driver focus']
        if opt.train_crop == 'random':
            crop_method = MultiScaleRandomCrop(opt.scales_inside, opt.sample_size)
        elif opt.train_crop == 'corner':
            crop_method = MultiScaleCornerCrop(opt.scales_inside, opt.sample_size)
        elif opt.train_crop == 'center':
            crop_method = MultiScaleCornerCrop(
                opt.scales_inside, opt.sample_size, crop_positions=['c'])
        elif opt.train_crop == 'driver focus':
            crop_method = DriverFocusCrop(opt.scales_inside, opt.sample_size)
        train_spatial_transform = Compose([
            crop_method,
            MultiScaleRandomCrop(opt.scales_inside, opt.sample_size),
            ToTensor(opt.norm_value_inside), norm_method
        ])
        
        # 랜덤하게 프레임을 잘라서 시간적으로도 crop하여 시간적인 데이터 증강을한다.
        train_temporal_transform = UniformRandomSample(opt.sample_duration_inside, opt.end_second)
        train_target_transform = ClassLabel()
        train_horizontal_flip = RandomHorizontalFlip()
        training_data_inside = get_training_set_inside(opt, train_spatial_transform, train_horizontal_flip,
                                            train_temporal_transform, train_target_transform)
        train_loader = torch.utils.data.DataLoader(
            training_data_inside,
            batch_size=opt.batch_size,
            shuffle=True,
            num_workers=opt.n_threads,
            pin_memory=True)

    # outside 훈련 데이터 증강 및 데이터 로더 부분
    opt.scales_outside = [opt.initial_scale]
    for i in range(1, opt.n_scales_outside):
        opt.scales_outside.append(opt.scales_outside[-1] * opt.scale_step)
  
    opt.arch_outside = 'ConvLSTM'
    opt.mean_outside = get_mean(opt.norm_value_outside, dataset=opt.mean_dataset)
    opt.std_outside = get_std(opt.norm_value_outside)
    
    
    if opt.no_mean_norm and not opt.std_norm:
            norm_method = Normalize([0, 0, 0], [1, 1, 1])
    elif not opt.std_norm:
        norm_method = Normalize(opt.mean_outside, [1, 1, 1])
    else:
        norm_method = Normalize(opt.mean_outside, opt.std_outside)
        
        
        
    if not opt.no_train_inside:    
        assert opt.train_crop in ['random', 'corner', 'center', 'driver focus']
    if opt.train_crop == 'random':
        crop_method = MultiScaleRandomCrop(opt.scales_outside, opt.sample_size)
    elif opt.train_crop == 'corner':
        crop_method = MultiScaleCornerCrop(opt.scales_outside, opt.sample_size)
    elif opt.train_crop == 'center':
        crop_method = MultiScaleCornerCrop(
            opt.scales_outside, opt.sample_size, crop_positions=['c'])
    elif opt.train_crop == 'driver focus':
        crop_method = DriverFocusCrop(opt.scales_outside, opt.sample_size)
    train_spatial_transform = Compose([
        Scale(opt.sample_size),		
        ToTensor(opt.norm_value_outside) #, norm_method
    ])
    train_temporal_transform = UniformIntervalCrop(opt.sample_duration_outside, opt.interval)
    train_target_transform = Compose([
        Scale(opt.sample_size),
        ToTensor(opt.norm_value_outside)#, norm_method
    ])
    train_horizontal_flip = RandomHorizontalFlip()
    training_data_outside = get_training_set_outside(opt, train_spatial_transform, train_horizontal_flip,
                                        train_temporal_transform, train_target_transform)
    train_loader_outside = torch.utils.data.DataLoader(
            training_data_outside,
            batch_size=opt.batch_size,
            shuffle=True,
            num_workers=opt.n_threads,
            pin_memory=True)
    
    Conv_Block = Conv_Block().to(device)
    classifier = classifier().to(device)
    
    weights = [1, 2, 4, 2, 4]
    print(torch.cuda.is_available())
    class_weights = torch.FloatTensor(weights).cuda()
    criterion = nn.CrossEntropyLoss(weight=class_weights)
    optimizer = torch.optim.Adam(classifier.parameters(), lr=0.001)
    
    if not opt.no_cuda:
        criterion = criterion.cuda()
        
    inside_vectors = []
    outside_vectors = []
     
  
    min_length = min(len(train_loader), len(train_loader_outside))
    for i in range(min_length):
        inputs_in, targets_in = next(iter(train_loader))
        
        inside_vector = extract_inside_vector(inside_pth_path, inputs_in) 
        inside_vector = torch.from_numpy(inside_vector).float().to(device)
        inside_vectors.append(inside_vector)  
        
        if i % 10 == 0:
            print(f"progress: {(i/min_length)*100}%")
        
           
    for i in range(min_length):    
        inputs_out, targets_out = next(iter(train_loader_outside)) 
        
        outside_vector = extract_outside_vector(outside_pth_path, inputs_out)
        outside_vector = torch.from_numpy(outside_vector).float().to(device)
        outside_vector = Conv_Block(outside_vector)
        outside_vectors.append(outside_vector)

        if i % 10 == 0:
            print(f"progress: {(i/min_length)*100}%")
            
       
    for epoch in range(opt.n_epochs + 1):    
        avg_acc = []
        for i in range(min_length):
            inputs_in, targets_in = next(iter(train_loader))
            combined_vector = torch.cat((inside_vectors[i], outside_vectors[i]), dim=1)
                    # print(combined_vector.shape)
            output = classifier(combined_vector)
            targets_in = targets_in.to(device)
            loss = criterion(output, targets_in)
            acc = calculate_accuracy(output, targets_in)
            
            #average of accuracy
            avg_acc.append(acc)
            
            
            if i % 10 == 0:
                avg_acc_value = sum(avg_acc) / len(avg_acc)
                print(f"Epoch [{epoch}/{opt.n_epochs}], Step [{i+1}/{len(train_loader)}], Loss: {loss.item()}, Accuracy: {acc}, avg_acc: {avg_acc_value}")

        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        optimizer.step()
            `
We implemented the code in this way, but the loss and accuracy are not satisfactory. Could there be any issues with how we structured the code?

**Comment by johook** [comment]
> @jihwan722 , please use a 3x3 conv layer with stride 3 to shrink the input size from 112x176 to 37x59 first. Code about classifier is not included in main_outside.py.

 After we use a 3x3 conv layer with stride 3 without pooling layer, we use 3x3 conv layer with stride 1 and 2x2 max pooling with stride 2. 
 Then we get the 17x28. It's different with the size in the paper 12x20.
 Please let me know the exact method. @yaorong0921

**Comment by yaorong0921** [comment]
Dear @johook, could you get 37x59 first after the 3x3 conv layer with the stride of 3? Do you mean that you could not get the dimension given in Table 2?

**Comment by yaorong0921** [comment]
Dear @jihwan722 , When fusing, the inside and outside videos are given into two different networks, which might take different numbers of frames. As given in Section IV-B (outside video), the input frames are all included in the time period before the second T using an interval L=5. In IV-C (in-cabin video), a 16-frame clip is given before the second T. We can understand how the classifier works as follows: each branch takes different frames before the second T to compute a feature; Then, two features are fused together, as described in Table 2.
